types/options.d.ts

```diff
@@ -52,48 +93,75 @@ export interface ComponentOptions<V extends Vue> {
   parent?: Vue;
   mixins?: (ComponentOptions<Vue> | typeof Vue)[];
   name?: string;
+  // TODO: support properly inferred 'extends'
   extends?: ComponentOptions<Vue> | typeof Vue;
   delimiters?: [string, string];
   comments?: boolean;
   inheritAttrs?: boolean;
 }
 
-export interface FunctionalComponentOptions {
+export interface FunctionalComponentOptions<Props = DefaultProps, PropDefs = PropsDefinition<Props>> {
   name?: string;
-  props?: string[] | { [key: string]: PropOptions | Constructor | Constructor[] };
+  props?: PropDefs;
   functional: boolean;
-  render(this: never, createElement: CreateElement, context: RenderContext): VNode | void;
+  render(this: undefined, createElement: CreateElement, context: RenderContext<Props>): VNode;
 }
 
-export interface RenderContext {
-  props: any;
+export interface RenderContext<Props=DefaultProps> {
+  props: Props;
   children: VNode[];
   slots(): any;
   data: VNodeData;
   parent: Vue;
   injections: any
 }
 
-export interface PropOptions {
-  type?: Constructor | Constructor[] | null;
+declare global {
+  interface StringConstructor {
+    '@@vueProp': string;
+  }
+  interface BooleanConstructor {
+    '@@vueProp': boolean;
+  }
+  interface NumberConstructor {
+    '@@vueProp': number;
+  }
+}
+
+export type Prop<T> = { '@@vueProp': T } | { new (...args: any[]): T & object }
```

Comment:
Instead of this global augmentation hack, the `Boolean`/`String`/`Number` constructors all already have call signatures that return `boolean`, `string`, and `number` respectively. So you could potentially write

```ts
export type Prop<T> = { (): T } | { new (...args: any[]): T & object }
```

Reply:
Indeed! It works!